singularity-eos/eos/eos_stiff.hpp

```diff
@@ -0,0 +1,200 @@
+// © 2021-2023. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract 89233218CNA000001
+// for Los Alamos National Laboratory (LANL), which is operated by Triad
+// National Security, LLC for the U.S.  Department of Energy/National
+// Nuclear Security Administration. All rights in the program are
+// reserved by Triad National Security, LLC, and the U.S. Department of
+// Energy/National Nuclear Security Administration. The Government is
+// granted for itself and others acting on its behalf a nonexclusive,
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_EOS_STIFF_HPP_
+#define _SINGULARITY_EOS_EOS_EOS_STIFF_HPP_
+
+// stdlib
+#include <cmath>
+#include <cstdio>
+#include <string>
+
+// Ports-of-call
+#include <ports-of-call/portability.hpp>
+
+// Base stuff
+#include <singularity-eos/base/constants.hpp>
+#include <singularity-eos/base/eos_error.hpp>
+#include <singularity-eos/base/robust_utils.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
+
+#define MYMAX(a, b) a > b ? a : b
+
+namespace singularity {
+
+using namespace eos_base;
+
+class StiffGas : public EosBase<StiffGas> {
+ public:
+  StiffGas() = default;
+  PORTABLE_INLINE_FUNCTION StiffGas( Real gm1,
+                                     Real Cv,
+                                     Real Pinf,
+                                     Real qq,
+                                     Real qp )
+      : _Cv(Cv), _gm1(gm1), _Pinf(Pinf), _qq(qq), _qp(qp),
+        _rho0( (_P0 + _Pinf) / (_gm1*_Cv*_T0) ),
+        _sie0( (_P0 + (_gm1+1.0)*_Pinf)/(_P0 + _Pinf) * _Cv*_T0 + _qq ),
+        _entr0( _Cv * log(_T0/( pow(_rho0*_gm1*_Cv, _gm1) ) + _qp )),
+        _bmod0( pow(_rho0,_gm1)*pow(_Cv,_gm1+1.0)*pow(_gm1,_gm1+2.0)*exp((_entr0-_qp)/_Cv) ),
+        _dpde0( _rho0*_gm1 ) {
+    checkParams();
+  }
+  StiffGas GetOnDevice() { return *this; }
+  PORTABLE_INLINE_FUNCTION Real TemperatureFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return MYMAX( 0.0, (rho*(sie-_qq)-_Pinf) / (rho*_Cv) );
+  }
+  PORTABLE_INLINE_FUNCTION void checkParams() const {
+    // Portable_require seems to do the opposite of what it should. Conditions
+    // reflect this and the code should be changed when ports-of-call changes
+    PORTABLE_ALWAYS_REQUIRE(_Cv >= 0, "Heat capacity must be positive");
+    PORTABLE_ALWAYS_REQUIRE(_gm1 >= 0, "Gruneisen parameter must be positive");
+  }
+  PORTABLE_INLINE_FUNCTION Real InternalEnergyFromDensityTemperature(
+      const Real rho, const Real temperature, Real *lambda = nullptr) const {
+    return MYMAX( 0.0, (rho*_Cv*temperature+_Pinf)/rho + _qq );
```

Comment:
you may want to replace `0` with `robust::small()` or something similar

Reply:
OK

Reply:
Recommend using the robust ratio function to accomplish the division.

Reply:
Corrected